phonon/lock.py

```diff
@@ -0,0 +1,26 @@
+import time
+
+import phonon
+import phonon.connections
+import phonon.exceptions
+
+
+class Lock(object):
+
+    def __init__(self, resource_key):
+        self.lock_key = "{}.lock".format(resource_key)
+        self.conn = phonon.connections.connection
+
+    def __enter__(self):
+        acquired = self.conn.client.setnx(self.lock_key, self.conn.id)
+        if not acquired:
+            connection_id = self.conn.client.get(self.lock_key)
+            if connection_id == self.conn.id:
```

Comment:
This bit actually makes the lock re-entrant, which is a new feature. In the same process we can acquire the lock twice without raising an exception, it just transparently yields the lock.